cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/SourceControlManagementHttpHandler.java

```diff
@@ -0,0 +1,111 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.gateway.handlers;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.BadRequestException;
+import io.cdap.cdap.common.RepositoryNotFoundException;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.security.AuditDetail;
+import io.cdap.cdap.common.security.AuditPolicy;
+import io.cdap.cdap.gateway.handlers.util.AbstractAppFabricHttpHandler;
+import io.cdap.cdap.internal.app.services.SourceControlManagementService;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfigRequest;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+/**
+ * {@link io.cdap.http.HttpHandler} for source control management.
+ */
+@Path(Constants.Gateway.API_VERSION_3 + "/namespaces/{namespace-id}/repository")
+public class SourceControlManagementHttpHandler extends AbstractAppFabricHttpHandler {
+  private final SourceControlManagementService sourceControlService;
+  private static final Gson GSON = new Gson();
+
+  @Inject
+  SourceControlManagementHttpHandler(SourceControlManagementService sourceControlService) {
+    this.sourceControlService = sourceControlService;
+  }
+
+  @PUT
+  @Path("/")
+  @AuditPolicy(AuditDetail.REQUEST_BODY)
+  public void setRepository(FullHttpRequest request, HttpResponder responder,
+                            @PathParam("namespace-id") String namespaceId) throws Exception {
+    RepositoryConfigRequest repoRequest = getRepository(request);
+    if (!repoRequest.shouldValidate()) {
+      if (!repoRequest.isValid()) {
+        throw new BadRequestException(String.format("Invalid repository configuration: %s.", repoRequest));
+      }
+      sourceControlService.setRepository(validateNamespaceId(namespaceId), repoRequest);
+      responder.sendString(HttpResponseStatus.OK, String.format("Updated repository configuration for namespace '%s'.",
+                                                                namespaceId));
+    }
+
+    // TODO: CDAP-20252 add the validate logic once the SourceControlManager module is ready
+  }
+
+  @GET
+  @Path("/")
+  public void getRepository(FullHttpRequest request, HttpResponder responder,
+                            @PathParam("namespace-id") String namespaceId) throws Exception {
+    NamespaceId namespace = validateNamespaceId(namespaceId);
+    RepositoryConfig repository = sourceControlService.getRepository(namespace);
+
+    if (repository == null) {
+      throw new RepositoryNotFoundException(namespace);
```

Comment:
we normally have the service throw the exception instead of putting the logic in the handler. We want to avoid putting business logic in the handler and have it responsible for only translating input/output.